internal/walk/walklinks.go

```diff
@@ -106,7 +106,7 @@ func filterPath(logger *log.Logger, path string, exclude []string, include []str
 }
 
 // Like filepath.WalkDir but resolves symlinks to directories.
-func walkDirWithLinks(ctx context.Context, root string, exclude []string, include []string, onlyThese []string, fn fs.WalkDirFunc) error {
+func walkDirWithLinks(ctx context.Context, root string, exclude []string, include []string, onlyThese []string, links bool, fn fs.WalkDirFunc) error {
 	logger := log.FromContext(ctx)
 
 	var walkFunc fs.WalkDirFunc
```

Comment:
Unclear why but when I run exodus-rsync locally, it panics.
```panic: runtime error: invalid memory address or nil pointer dereference```
The error points at line 124 (call to `filterPath()`) below and after some toying, I found that `d` (`fs.DirEntry`) is `<nil>` and so `d.IsDir()` is invalid.

This issue isn't specific to this PR, it's present on main branch too. I'm just bringing this up to ask for help with testing. A solution to this odd problem would also be great if anyone has experienced something similar.

Reply:
What do you need to do to reproduce it?

I'd suspect it's a bug caused by us checking `err` too late.  I believe the way some of these callback functions work is: if you receive a non-nil error, then the arguments such as the DirEntry (or FileInfo in the similar filepath.Walk case) should not be used.

In this function we're not checking `err` until *after* some code which tries to access `d`, so that's probably a bug on our end and the `err` check should be moved earlier, at least above the `filterPath` call.

If the crash still happens after doing that, then this theory is wrong.

Reply:
That was it! Moved the check up.